src/DeviceInput/deviceInputSystem.ts

```diff
@@ -0,0 +1,310 @@
+                    }
+                }
+
+                const touchButton = this._inputs[`${DeviceInputSystem.TOUCH_DEVICE}-${this._pointerIds.lastIndexOf(evt.pointerId)}`];
+                if (touchButton) {
+                    touchButton[0] = evt.clientX;
+                    touchButton[1] = evt.clientY;
+                    touchButton[2] = 1;
+                    this._activeTouchNumber++;
+                }
+            }
+        });
+
+        this._pointerUpEvent = ((evt) => {
+            if (evt.pointerType == "mouse") {
+                const mouseButton = this._inputs[DeviceInputSystem.MOUSE_DEVICE];
+                if (mouseButton) {
+                    mouseButton[evt.button + 2] = 0;
+                }
+            }
+            else if (evt.pointerType == "touch") {
+                const touchIndex = this._pointerIds.lastIndexOf(evt.pointerId);
+                const touchButton = this._inputs[`${DeviceInputSystem.TOUCH_DEVICE}-${touchIndex}`];
+                if (touchButton) {
+                    this._pointerIds.splice(touchIndex, 1);
+
+                    // Push values of touch inputs down
+                    for (let i = touchIndex; i < this._activeTouchNumber; i++) {
+                        const nextTouch = this._inputs[`${DeviceInputSystem.TOUCH_DEVICE}-${i + 1}`];
+                        const currentTouch = this._inputs[`${DeviceInputSystem.TOUCH_DEVICE}-${i}`];
+
+                        if (currentTouch && nextTouch) {
+                            currentTouch[0] = nextTouch[0];
+                            currentTouch[1] = nextTouch[1];
+                            currentTouch[2] = nextTouch[2];
+                        }
+                        else if (currentTouch) {
+                            currentTouch[0] = 0;
+                            currentTouch[1] = 0;
+                            currentTouch[2] = 0;
+                        }
+                    }
+
+                    --this._activeTouchNumber;
+                }
+            }
+        });
+
+        this._elementToAttachTo.addEventListener("pointermove", this._pointerMoveEvent);
+        this._elementToAttachTo.addEventListener("pointerdown", this._pointerDownEvent);
+        this._elementToAttachTo.addEventListener("pointerup", this._pointerUpEvent);
+    }
+
+    /**
+     * Handle all actions that come from gamepad interaction
+     */
+    private _handleGamepadActions() {
+        this._gamepadConnectedEvent = ((evt: any) => {
+            const deviceName = `${evt.gamepad.id}-${evt.gamepad.index}`;
+            this._registerDevice(deviceName, evt.gamepad.buttons.length + evt.gamepad.axes.length);
+        });
+
+        this._gamepadDisconnectedEvent = ((evt: any) => {
+            const deviceName = `${evt.gamepad.id}-${evt.gamepad.index}`;
+            this._unregisterDevice(deviceName);
+        });
+
+        window.addEventListener("gamepadconnected", this._gamepadConnectedEvent);

```

Comment:
I would recommend to implement IDisposable and unwire all the events there

Reply:
I had a similar comment earlier and @PolygonalSun added a Dispose method that removes the event listeners. Probably should add the `extends IDiposable` to the class declaration too.

Reply:
implemented IDisposable